elements/lisk-tree/src/merkle_tree.ts

```diff
@@ -163,10 +154,219 @@ export class MerkleTree {
+
+					// If there is only one node in the current layer
+					if (currentLayer.length === 1) {
+						// Find the next lower layer with odd number of nodes
+						let currentLayerIndex = currentNode.layerIndex - 1;
+						while (
+							currentLayerIndex >= 0 &&
+							treeStructure[currentLayerIndex].length % 2 === 0
+						) {
+							currentLayerIndex -= 1;
+						}
+						currentLayer = treeStructure[currentLayerIndex];
+						const lastNodeOfLowerOddLayer = {
+							direction: NodeSide.RIGHT,
+							hash: currentLayer[currentLayer.length - 1].hash,
+						};
+						path.push(lastNodeOfLowerOddLayer);
+						const parentNodeHash = hash(
+							Buffer.concat(
+								[BRANCH_PREFIX, currentNode.hash, lastNodeOfLowerOddLayer.hash],
+								BRANCH_PREFIX.length +
+									currentNode.hash.length +
+									lastNodeOfLowerOddLayer.hash.length,
+							),
+						);
+						currentNode = this.getNode(parentNodeHash);
+					}
+					// If there is more than one node and the current node is not the last node in the layer
+					else if (
+						currentNode.hash.compare(
+							currentLayer[currentLayer.length - 1].hash,
+						) !== 0
+					) {
+						const pairInfo =
+							currentNode.nodeIndex % 2 === 0
+								? {
+										direction: NodeSide.RIGHT,
+										hash: currentLayer[currentNode.nodeIndex + 1].hash,
+								  }
+								: {
+										direction: NodeSide.LEFT,
+										hash: currentLayer[currentNode.nodeIndex - 1].hash,
+								  };
+						path.push(pairInfo);
+						const leftHashBuffer =
+							pairInfo.direction === NodeSide.LEFT
+								? pairInfo.hash
+								: currentNode.hash;
+						const rightHashBuffer =
+							pairInfo.direction === NodeSide.LEFT
+								? currentNode.hash
+								: pairInfo.hash;
+						const parentNodeHash = hash(
+							Buffer.concat(
+								[BRANCH_PREFIX, leftHashBuffer, rightHashBuffer],
+								BRANCH_PREFIX.length +
+									leftHashBuffer.length +
+									rightHashBuffer.length,
+							),
+						);
+						currentNode = this.getNode(parentNodeHash);
+					}
+					// If current layer has more than one node
+					else {
+						// Find the next higher layer with odd number of nodes
+						let currentUpperLayerIndex = currentNode.layerIndex + 1;
+						while (
+							currentUpperLayerIndex < this._getHeight() &&
+							treeStructure[currentUpperLayerIndex].length % 2 === 0
+						) {
+							currentUpperLayerIndex += 1;
+						}
+						currentLayer = treeStructure[currentUpperLayerIndex];
+						// Pair the last node in the layer
+						const pairInfo = {
+							direction: NodeSide.LEFT,
+							hash: currentLayer[currentLayer.length - 1].hash,
+						};
+						const parentNodeHash = hash(
+							Buffer.concat(
+								[BRANCH_PREFIX, pairInfo.hash, currentNode.hash],
+								BRANCH_PREFIX.length +
+									pairInfo.hash.length +
+									currentNode.hash.length,
+							),
+						);
+						if (parentNodeHash.toString('binary') in this._hashToValueMap) {
```

Comment:
```suggestion
						if (this._hashToValueMap[parentNodeHash.toString('binary')] !== undefined) {
```
Im not sure if this is better, but it is done in this way in other places

Reply:
To me the first way looks cleaner